pkg/server/v3/delta_test.go

```diff
@@ -0,0 +1,413 @@
+	nonce     int
+	sendError bool
+	grpc.ServerStream
+}
+
+func (stream *mockDeltaStream) Context() context.Context {
+	return stream.ctx
+}
+
+func (stream *mockDeltaStream) Send(resp *discovery.DeltaDiscoveryResponse) error {
+	// check that nonce is monotonically incrementing
+	stream.nonce = stream.nonce + 1
+	if resp.Nonce != fmt.Sprintf("%d", stream.nonce) {
+		stream.t.Errorf("Nonce => got %q, want %d", resp.Nonce, stream.nonce)
+	}
+	// check resources are non-empty
+	if len(resp.Resources) == 0 {
+		stream.t.Error("Resources => got none, want non-empty")
+	}
+	// check that type URL matches in resources
+	if resp.TypeUrl == "" {
+		stream.t.Error("TypeUrl => got none, want non-empty")
+	}
+
+	for _, res := range resp.Resources {
+		if res.Resource.TypeUrl != resp.TypeUrl {
+			stream.t.Errorf("TypeUrl => got %q, want %q", res.Resource.TypeUrl, resp.TypeUrl)
+		}
+	}
+
+	stream.sent <- resp
+	if stream.sendError {
+		return errors.New("send error")
+	}
+	return nil
+}
+
+func (stream *mockDeltaStream) Recv() (*discovery.DeltaDiscoveryRequest, error) {
+	req, more := <-stream.recv
+	if !more {
+		return nil, errors.New("empty")
+	}
+	return req, nil
+}
+
+func makeMockDeltaStream(t *testing.T) *mockDeltaStream {
+	return &mockDeltaStream{
+		t:    t,
+		ctx:  context.Background(),
+		sent: make(chan *discovery.DeltaDiscoveryResponse, 10),
+		recv: make(chan *discovery.DeltaDiscoveryRequest, 10),
+	}
+}
+
+func makeDeltaResponses() map[string][]cache.DeltaResponse {
+	return map[string][]cache.DeltaResponse{
+		rsrc.EndpointType: {
+			&cache.RawDeltaResponse{
+				Resources:         []types.Resource{endpoint},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: rsrc.EndpointType},
+				SystemVersionInfo: "1",
+			},
+		},
+		rsrc.ClusterType: {
+			&cache.RawDeltaResponse{
+				Resources:         []types.Resource{cluster},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: rsrc.ClusterType},
+				SystemVersionInfo: "2",
+			},
+		},
+		rsrc.RouteType: {
+			&cache.RawDeltaResponse{
+				Resources:         []types.Resource{route},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: rsrc.RouteType},
+				SystemVersionInfo: "3",
+			},
+		},
+		rsrc.ListenerType: {
+			&cache.RawDeltaResponse{
+				Resources:         []types.Resource{listener},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: rsrc.ListenerType},
+				SystemVersionInfo: "4",
+			},
+		},
+		rsrc.SecretType: {
+			&cache.RawDeltaResponse{
+				SystemVersionInfo: "5",
+				Resources:         []types.Resource{secret},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: rsrc.SecretType},
+			},
+		},
+		rsrc.RuntimeType: {
+			&cache.RawDeltaResponse{
+				SystemVersionInfo: "6",
+				Resources:         []types.Resource{runtime},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: rsrc.RuntimeType},
+			},
+		},
+		// Pass-through type (xDS does not exist for this type)
```

Comment:
`xDS does not exist for this type` seems a bit odd. Maybe `types without explicit handling`?

Reply:
Yea I like that better, makes it easier to understand